### ISSUE TITLE
Support authentication using OpenID Connect

### ISSUE THREAD

**Issue by karimhm** [processone/ejabberd#3437]
**Is your feature request related to a problem? Please describe.**
To allow users to use a single account for all the services or to use third-party identity providers. [SSO](https://wikipedia.org/wiki/Single_sign-on) is needed.

**Describe the solution you'd like**
Add support for [OpenID Connect](https://openid.net/connect). `OpenID Connect` uses [JWT](https://en.wikipedia.org/wiki/JSON_Web_Token) as the token format and is compatible with the existing ejabberd [JWT Authentication](https://docs.ejabberd.im/admin/configuration/authentication/#jwt-authentication) module.

**Describe alternatives you've considered**
Implement a sidecar service that downloads the `OpenID Connect` JWK Keys locally and perform a rolling restart so ejabberd would load the new JWK Key.

**Additional context**
ejabberd [JWT Authentication](https://docs.ejabberd.im/admin/configuration/authentication/#jwt-authentication) module grabs the `JWK Key` locally via the [jwt_key](https://docs.ejabberd.im/admin/configuration/toplevel/#jwt-key) option.

I would suggest adding a new option to ejabberd [JWT Authentication](https://docs.ejabberd.im/admin/configuration/authentication/#jwt-authentication) module such as `jwt_oidc_jwks_uri` that will contain the `OpenID Connect` URL for the Client's JSON Web Key Set or an option such as `jwt_oidc_issuer` that will point to the server which provides the `OpenID Connect` configuration.

**Comment by karimhm** [comment]
A certified Erlang library already exists: [oidcc](https://github.com/indigo-dc/oidcc).

**Comment by Robbilie** [comment]
it would also be nice to just be able to work with jwts that dont have a specific jid field. this would allow for using existing tokens from something like keycloak or auth0. ie. define seperate claims for user and server definition, sub as the user claim and maybe a static claim for the server…

i would love to see jwks integration, maybe reloading would be nice too, im currently considering downloading a jwks at the start of the container and write the jwks somewhere where the ejabberd.yml is configured to read it from, no reloading though…

**Comment by debdolph** [comment]
Any progress on this?

**Comment by kir-repnikov** [comment]
I recommend to reconsider status and priority for this feature request. Nowadays key rotation is critical aspect of security. 

Certainly, we can use workaround described by author. But it might be really great improvement.

**Comment by xundeenergie** [comment]
Oh yes.
OIDC-Support for XMPP would be great!!

**Comment by licaon-kter** [comment]
@xundeenergie look at the labels, was not completed but put aside...

**Comment by xundeenergie** [comment]
> @xundeenergie look at the labels, was not completed but put aside...

ok... so we have to wait for.

**Comment by licaon-kter** [comment]
I would not hold my breath :)

**Comment by Alexei-Barnes** [comment]
Not supporting a changeable JWK in <current year> is crazy, how does anyone expect this to work with, e.g. Auth0, or any other major identity provider?

**Comment by Robbilie** [comment]
Going into the issue section of an open source project with that kind of attitude in <current year> is crazy...

**Comment by Dunedan** [comment]
I'd be interested in OpenID Connect support as well to be able to use ejabberd as identity provider.

**Comment by DaDummy** [comment]
Based on this issue being closed I do wonder... Is there a specific reason for why support for OIDC is currently not wanted?

**Comment by badlop** [comment]
I closed this issue when more than a year and half had passed with it opened and nobody had been able to work on it.

The topic it refers to is relevant (as seen in the comments), so I tagged the issue as "To reconsider" and assigned it to the "Parking Lot", expressing that the feature is wanted... but until now, nobody made a step up in a very reasonable time of 22 months.

There are many demonstrations of interest in the feature, unfortunately all this interest only surfaces as messages in the issue, not in any practical steps that could help advance in the feature itself.
